File: 19 09 25/permutationsequence.py
# LeetCode 60: Permutation Sequence
# Explanation:
# 1. The number of permutations for n elements is n!.
# 2. Fix the first digit by determining how many permutations are in each block of size (n-1)!.
# 3. Use k to find which block to choose and remove that element.
# 4. Repeat for the remaining numbers.
# Time Complexity: O(n^2) (because list removals are O(n) and we do it n times)
# Space Complexity: O(n)

class Solution:
    def getPermutation(self, n: int, k: int) -> str:
        import math
        nums = [str(i) for i in range(1, n+1)]
        k -= 1
        result = []
        for i in range(n, 0, -1):
            f = math.factorial(i-1)
            index = k // f
            result.append(nums.pop(index))
            k %= f
        return ''.join(result)

# Example usage:
# sol = Solution()
# print(sol.getPermutation(3, 3))  # "213"
# print(sol.getPermutation(4, 9))  # "2314"

File: 19 09 25/permutationsequence.js
// LeetCode 60: Permutation Sequence
// Explanation:
// 1. Use factorial blocks to find which number to choose at each step
// 2. Remove chosen numbers from the list as we build the result
// Time Complexity: O(n^2)
// Space Complexity: O(n)

var getPermutation = function(n, k) {
    const fact = [1];
    for (let i = 1; i <= n; i++) fact[i] = fact[i-1]*i;
    const nums = [];
    for (let i = 1; i <= n; i++) nums.push(i.toString());
    k--;
    let res = '';
    for (let i = n; i >= 1; i--) {
        const idx = Math.floor(k / fact[i-1]);
        res += nums[idx];
        nums.splice(idx, 1);
        k %= fact[i-1];
    }
    return res;
};

